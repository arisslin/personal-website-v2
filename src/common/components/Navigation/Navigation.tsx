import { Link } from 'gatsby';
import * as React from 'react';
import styled from 'styled-components/macro';
import { Link as LinkType } from '../../../types/dataTypes';

export type NavigationProps = {
  links: LinkType[];
};

const Navigation = ({ links }: NavigationProps) => {
  return (
    <NavStyled data-testid={testId}>
      <div className='layout-container'>
        {links.map((link, index) => (
          <Link
            key={'nav-link-' + index}
            to={link.href}
            activeClassName='active'
          >
            {link.text}
          </Link>
        ))}
      </div>
    </NavStyled>
  );
};

const NavStyled = styled.nav`
  position: sticky;
  top: 0;
  height: 40px;
  background: var(--color-primary);
  z-index: 1;

  .layout-container {
    display: flex;
    align-items: stretch;
    height: 100%;
  }

  a {
    display: flex;
    align-items: center;
    padding: 0 8px;
    color: white;
    text-decoration: none;
  }

  a.active {
    background: var(--color-light);
    color: var(--color-dark);
  }
`;

export const testId = 'navigation';

export default Navigation;
