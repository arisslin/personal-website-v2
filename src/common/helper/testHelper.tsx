import { cleanup, fireEvent, render } from '@testing-library/react';

/**
 * Does a snapshot test for component it got.
 * @param {JSX.Element} component to creating snapshot for
 * @param {string} variant to create a postfix for the displayed test string
 * @return {void}
 */
export function doSnapshotTest(component: JSX.Element, variant?: string): void {
  const postfix = `${component.type.name as string} ${variant ? variant : ''}`;

  test(`Snapshot test for ${postfix}`, () => {
    const { container } = render(component);

    expect(container).toMatchSnapshot();
    cleanup();
  });
}

/**
 * Fires an click event with 'fireEvent' from '@testing-library/react' on an
 * element.
 * @param {HTMLElement} element to fire click on.
 * @param {boolean} bubbles attribute of 'MouseEvent'.
 * @param {boolean} cancelable attribute of 'MouseEvent'.
 * @return {void}
 */
export function fireClickEvent(
  element: Element,
  bubbles: boolean,
  cancelable: boolean
) {
  fireEvent(element, new MouseEvent('click', { bubbles, cancelable }));
}
