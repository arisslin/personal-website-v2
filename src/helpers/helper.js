export const getText = (wantText) => {
  if (typeof wantText === 'boolean') {
    return wantText ? 'text' : null;
  }

  return null;
};
