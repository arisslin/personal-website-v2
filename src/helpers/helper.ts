export const getText = (wantText: boolean) => (wantText ? 'text' : null);
