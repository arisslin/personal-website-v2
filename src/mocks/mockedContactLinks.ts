import { ContactLinkProps } from '../common/components/ContactLink/ContactLink';
import { FaEnvelope, FaGithub, FaGlobe, FaXing } from 'react-icons/fa';

export const mockedContactLinks: ContactLinkProps[] = [
  {
    text: 'gueteklasse-a.de',
    href: 'https://www.gueteklasse-a.de',
    icon: FaGlobe,
  },
  {
    text: 'andreas.rissling@gmx.net',
    href: 'mailto:andreas.rissling@gmx.net',
    icon: FaEnvelope,
  },
  {
    text: 'GitHub',
    href: 'https://github.com/arisslin',
    icon: FaGithub,
  },
  {
    text: 'Xing',
    href: 'https://www.xing.com/profile/Andreas_Rissling3/',
    icon: FaXing,
  },
];
