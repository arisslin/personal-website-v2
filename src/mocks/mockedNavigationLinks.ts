import { Link } from '../types/dataTypes';

export const mockedNavigationLinks: Link[] = [
  { href: '', text: 'Home' },
  {
    href: '/aboutme',
    text: 'About me',
  },
];
