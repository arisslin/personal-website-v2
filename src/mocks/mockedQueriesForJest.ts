import * as Gatsby from 'gatsby';

export function mockQuerySiteMetaData(): void {
  const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');

  useStaticQuery.mockImplementation(() => ({
    site: {
      siteMetadata: {
        description: 'My description',
      },
    },
  }));
}
