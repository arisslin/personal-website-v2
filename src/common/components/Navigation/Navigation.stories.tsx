import * as React from 'react';
import { Meta, Story } from '@storybook/react';
import NavigationComponent, { NavigationProps } from './Navigation';
import { mockedNavigationLinks } from '../../../mocks/mockedNavigationLinks';

export default {
  title: 'Components/Navigation',
  component: NavigationComponent,
  parameters: {
    docs: {
      description: {
        component: 'Follows',
      },
    },
  },
} as Meta;

export const navigation: Story<NavigationProps> = (args) => (
  <NavigationComponent links={args.links} />
);
navigation.args = {
  links: mockedNavigationLinks,
};
