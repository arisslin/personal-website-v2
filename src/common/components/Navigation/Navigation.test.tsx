import * as React from 'react';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import Navigation from './Navigation';
import { mockedNavigationLinks } from '../../../mocks/mockedNavigationLinks';
import { doSnapshotTest, fireClickEvent } from '../../helper/testHelper';

describe('Navigation', () => {
  const component = <Navigation links={mockedNavigationLinks} />;

  afterEach(() => {
    cleanup();
  });

  it('renders with empty links', () => {
    render(<Navigation links={[]} />);

    expect(screen.getByRole('navigation')).toBeInTheDocument();
  });

  it('renders with links', () => {
    render(component);

    mockedNavigationLinks.forEach((link) => {
      const renderedLink = screen.getByText(link.text);

      expect(renderedLink).toBeInTheDocument();
      expect(renderedLink.getAttribute('href')).toBe(link.href);
    });
  });

  it('changes it style on menu icon click', () => {
    render(component);

    const classNameBeforeClick = screen.getByRole('navigation').className;
    const menuIcon = screen.getByRole('navigation').querySelector('svg');

    menuIcon && fireClickEvent(menuIcon, true, true);

    const classNameAfterClick = screen.getByRole('navigation').className;

    expect(classNameBeforeClick).not.toBe(classNameAfterClick);
  });

  doSnapshotTest(component);
});
