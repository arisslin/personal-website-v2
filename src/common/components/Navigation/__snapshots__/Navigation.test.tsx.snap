// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Snapshot test for Navigation  1`] = `
<div>
  <nav
    class="Navigation__NavStyled-sc-1jart14-0 hWrVYh"
  >
    <div
      class="layout-container"
    >
      <a
        href="/"
      >
        Home
      </a>
      <a
        href="/aboutme"
      >
        Ãœber mich
      </a>
      <a
        href="/contact"
      >
        Kontakt
      </a>
      <svg
        class="navigation__menu-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </div>
  </nav>
</div>
`;
